[RDY] vim-patch:8.0.1738: ":args" output is hard to read

blueyed opened the pull request:
Problem:    ":args" output is hard to read.
Solution:   Make columns with the names if the output is more than one line.
https://github.com/vim/vim/commit/5d69da462f584a3aefb3427b127334bf9af3a4b0

Likely a bit wrong (or missing a follow-up patch), at least I am seeing this with `:args`:
```
[vim-1ebff3dc9.patch] vim-8.1.0114.patch vim-8.1.0429.patch vim-8.1.1365.patch
vim-1fbfe7c48.patch vim-8.1.0200.patch vim-8.1.0434.patch vim-58f60ca2f.patch
...
vim-8.0.1339.patch vim-8.1.0362.patch vim-8.1.0619.patch vim-83799a7b7.patch
vim-8.0.1557.patch vim-8.1.0404.patch vim-8.1.1017.patch
vim-8.0.1738.patch vim-8.1.0410.patch vim-8.1.1245.patch
See ":help feature-compile"


Press ENTER or type command to continue
```

While it looks like this with Vim:
```
[vim-1ebff3dc9.patch] vim-8.1.0114.patch    vim-8.1.0429.patch    vim-8.1.1365.patch
vim-1fbfe7c48.patch   vim-8.1.0200.patch    vim-8.1.0434.patch    vim-58f60ca2f.patch
...
vim-8.0.1339.patch    vim-8.1.0362.patch    vim-8.1.0619.patch    vim-83799a7b7.patch
vim-8.0.1557.patch    vim-8.1.0404.patch    vim-8.1.1017.patch
vim-8.0.1738.patch    vim-8.1.0410.patch    vim-8.1.1245.patch

Press ENTER or type command to continue
```

The "See ":help feature-compile"" is very odd there (and the extra newlines).

blueyed committed:
vim-patch:8.0.1738: ":args" output is hard to read

Problem:    ":args" output is hard to read.
Solution:   Make columns with the names if the output is more than one line.
https://github.com/vim/vim/commit/5d69da462f584a3aefb3427b127334bf9af3a4b0

vim-patch:8.0.1740: warning for signed-unsigned incompatibility

Problem:    Warning for signed-unsigned incompatibility.
Solution:   Change type from "char *" to "char_u *". (John Marriott)
https://github.com/vim/vim/commit/405dadb63ea2b7aa4c8c659807506a35a8a9504c

Removes ported legacy test that was re-added later.
Ref: https://github.com/neovim/neovim/pull/10147#issuecomment-512609513

blueyed commented on src/nvim/ex_cmds2.c:
```diff
@@ -1791,18 +1791,17 @@ void ex_args(exarg_T *eap)
-      for (int i = 0; i < ARGCOUNT; i++) {
-        if (i == curwin->w_arg_idx) {
-          msg_putchar('[');
+      char **items = (char **)xmalloc(sizeof(char *) * ARGCOUNT);
```
```
../src/nvim/ex_cmds2.c: In function ‘ex_args’:
../src/nvim/ex_cmds2.c:1794:54: error: conversion to ‘long unsigned int’ from ‘int’ may change the sign of the result [-Werror=sign-conversion]
       char **items = (char **)xmalloc(sizeof(char *) * ARGCOUNT);
                                                      ^
../src/nvim/ex_cmds2.c:1803:9: error: implicit declaration of function ‘list_in_columns’ [-Werror=implicit-function-declaration]
         list_in_columns(items, ARGCOUNT, curwin->w_arg_idx);
         ^
```

What to do here?

justinmk replied:
I think we usually do something like:

```
assert(ARGCOUNT >= 0);
char **items = xmalloc(sizeof(char *) * (size_t)ARGCOUNT);
```

Note that we don't cast `xmalloc()` result.

blueyed replied:
Adjusted.

blueyed commented:
Came to it via vim/vim@e961cba3c (tag: v8.1.0404).

janlazo commented:
`test_arglist.vim` is disabled and it fails because of 8.1.0493.

janlazo commented:
https://github.com/neovim/neovim/pull/10245 should unblock this.

blueyed commented:
Rebased, fixed.

Still fails though:

> Test_argument line 28: Expected 'a   b   [c] d' but got 'a   b   [c] d   \nSee ":help feature-compile"'

Not really clear if we want that here?! - likely not.  /cc @justinmk

blueyed commented:
Hangs at https://github.com/blueyed/neovim/blob/afc6273cd10b88bcdddbbe3aa43a0ab2495ca433/test/functional/legacy/arglist_spec.lua#L117-L119.

Using `command('let result = execute("args")')` works around it - but this might indicate something, e.g. some unexpected hit-enter prompt?!

justinmk commented:
> Using `command('let result = execute("args")')` works around it - but this might indicate something, e.g. some unexpected hit-enter prompt?!

Yes, `:args` could potentially cause hit-enter.

blueyed commented:
Not really for the issue at hand though?

justinmk commented:
`screen:expect()` or `screen:snapshot_debug()` would show the actual screen contents.

blueyed commented:
> screen:snapshot_debug()

There's no screen there (easily) is there?

Anyway, we have the test in oldtests: https://github.com/neovim/neovim/blob/afc6273cd10b88bcdddbbe3aa43a0ab2495ca433/src/nvim/testdir/test_arglist.vim#L100 - so I suggest to just remove it here?  (and use the one from Vim)

blueyed commented:
It was re-enabled/used in 77bfbc300.

blueyed commented:
This test is equal, but Vim has two lines more: https://github.com/neovim/neovim/blob/afc6273cd10b88bcdddbbe3aa43a0ab2495ca433/src/nvim/testdir/test_arglist.vim#L167-L168

blueyed commented:
Turns out that there was a hit-enter prompt actually - a problem with the original patch.
Waiting for https://github.com/vim/vim/pull/4715 and will incorporate it here then.

blueyed committed:
vim-patch:8.1.1737: :args command that outputs one line gives more prompt

Problem:    :args command that outputs one line gives more prompt.
Solution:   Only output line break if needed. (Daniel Hahler, closes vim/vim#4715)
https://github.com/vim/vim/commit/949f1989cba8bf7653316c2b1444c26f1536bfab

janlazo commented on src/nvim/version.c:
```diff
@@ -2076,6 +2076,7 @@ void list_in_columns(char_u **items, int size, int current)
   int nrow = item_count / ncol + (item_count % ncol ? 1 : 0);
+  int cur_row = 1;
 
   // i counts columns then rows.  idx counts rows then columns.
```
The patch has ` "i" counts`. `idx` should be double-quoted for consistency.

blueyed replied:
Thanks for catching!
I'll amend it when merging it locally.

blueyed replied:
> `idx` should be double-quoted for consistency.

Felt a bit uneasy about it, since it is unnecessary diff from Vim, but since it requested linting for `++i` => `i++` then anyway.. :man_shrugging:
